fix Newton compressible terms.

MFraters opened the pull request:
Investigating issue #4388 I found that the compressible terms for the Newton (and therefore the defect correction Picard solver) are not added in the correct way. This lead to the solver not converging (even diverging). The test case proveded by @elodie-kendall does not converge for me and the solutions for the first timestep (0) look now identical to me. 

@elodie-kendall, could you check whether this solves the issue for you?

gassmoeller commented on source/simulator/assemblers/newton_stokes.cc:
```diff
@@ -588,7 +588,7 @@ namespace aspect
                 for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
                   {
-                    data.local_matrix(i,j) += (- eta_two_thirds * (scratch.div_phi_u[i] * scratch.div_phi_u[j])) * JxW;
+                    data.local_matrix(i,j) -= (- eta_two_thirds * (scratch.div_phi_u[i] * scratch.div_phi_u[j])) * JxW;
```
@MFraters I find this change confusing. The old version is identical to the term in the normal Stokes equation (e.g. stokes.cc:532). Does this term have a different sign for the Newton matrix?

I wonder if the convergence issues are instead related to #1967? I just checked and it seems there is no compressibility correction (i.e. the term in this code) for the Newton Stokes preconditioner, which could negatively affect convergence behavior.

MFraters commented:
@gassmoeller Thanks for your review. It made me realize that although the fix kinda worked it was not correct (the maximum velocity where not the same and I should have been more suspicious of that it took a lot more iterations for the dc to converge). 

The issue doesn't have to do with the preconditioner, since that would only affect the linear convergence, which is not the issue here. The issue is the defect correction rhs. A normal system is Ax=b and a DC system is A dx = - (Ax-b) = -Ax + b. In the compressible case the compressible terms where not added to the rhs (they are part of A, so they need to be substracted from the rhs). 

This solution make sure that the maximum velocities are now the same and the number of iterations are the same (+-1). I need to clean it up and update the tests (also adding a stokes version for comparison)

elodie-kendall commented:
Thanks a lot, can I use this branch now?

MFraters commented:
@elodie-kendall, yes, please try out this branch and see if it fixes the issues for you.

gassmoeller commented:
@mfraters one test (`newton_compressible`) does not converge with this change. Is this because the change is not correct, or is it a problem with the test? https://jenkins.tjhei.info/blue/organizations/jenkins/aspect/detail/PR-4422/7/tests

MFraters commented:
A bit of both, but mostly the test. I changed the tests locally so that the test with the name newton actually uses the full newton solver instead of the dc Picard (I misnamed it when writing the first test for this pull request). I also added tests for the full Picard and the dc Picard (but haven't added them in a commit yet). The full stokes and dc Picard tests work fine. The Newton has the issue that the Newton solver is not stabilized for compressible: there are extra negative terms which break the positive definite condition in this setup.  That is why the Newton solver fails. 

I am doubting whether to invest time to find a compressible test case which can converge, but that is going to take time and I think is a little bit out of scope of this pull request. So I am thinking of just removing the Newton test and stick with the full Stokes and dc Picard versions. Because the dc Picard uses the same solver as the Newton solver, fixing the dc Picard solver also fixes the Newton solver, even though the linear solver can't converge for this problem with the Newton solver.

So to summarize: I think this fixes both the DC Picard and Newton solver for compressible. The Newton solver for compressible is not stable for this test case and therefore fails.

MFraters commented:
This pull request is ready again for review. I have added two tests and removed one. The tests I added compare the normal Picard iteration to the DC Picard iteration and show that the results are the same. I removed the Netwon version due to the stabilization issues and since this fix is independent to whether derivatives are added in the assembly.

@elodie-kendall, could you report whether this has solved the issue for you?

elodie-kendall commented:
Hi Menno, thanks for adding these latest changes and tests. However, the defect correction stokes is still not converging for a case which works with the simpler "iterated Stokes". I attach the prm as an example
```
#!/usr/bin/env python3

####  Global parameters
set Dimension                              = 2
set Use years in output instead of seconds = true
set Start time                             = 0
set End time                               = 700.e6  
set Output directory                       = output
set Nonlinear solver scheme                = single Advection, iterated defect correction Stokes
set Nonlinear solver tolerance             = 5e-4
set Max nonlinear iterations               = 150
set Max nonlinear iterations in pre-refinement = 0
set Pressure normalization                 = surface
set Maximum time step                      = 1e6
#set Number of cheap Stokes solver steps    = 200  # aspect 2.2
set Adiabatic surface temperature          = 1600 
set CFL number                             = 1

set Resume computation          = auto 
subsection Checkpointing
  set Steps between checkpoint  = 100
  set Time between checkpoint   = 0
end
subsection Termination criteria
  set Checkpoint on termination           = true
  set End step                            = 1
  set Termination criteria                = end time #end step
end

subsection Solver parameters
  subsection Stokes solver parameters
    set GMRES solver restart length = 250
    set Linear solver tolerance             = 1e-5
    set Maximum number of expensive Stokes solver steps = 1000
    set Number of cheap Stokes solver steps    = 200 
  end
end

#### Model Geometry
subsection Geometry model
  set Model name = spherical shell

  subsection Spherical shell
    set Inner radius  = 3485e3
    set Outer radius  = 6371e3 #5961e3
    set Opening angle = 360
  end
end

#### Mesh refinement
subsection Mesh refinement
#  set Coarsening fraction = 0.2
#  set Refinement fraction = 0.8
  set Initial adaptive refinement        = 3
  set Initial global refinement          = 5
  set Time steps between mesh refinement = 0
  set Strategy                           = minimum refinement function
  subsection Minimum refinement function
    set Variable names      = depth,t
    set Coordinate system = depth
    set Function expression = if(depth<=200e3, 8,if(depth<=620e3, 7,if(depth<=680e3,6,if(depth>=2586e3,7,5))))
    #if(depth<560e3,0,if(depth>760e3,0,6))
#        if(depth<560e3,5,if(depth>760e3,5,6))
  end
end

subsection Boundary temperature model
  #set Allow fixed temperature on outflow boundaries = true  # default
  set Fixed temperature boundary indicators         = bottom, top
end
subsection Boundary velocity model
  set Tangential velocity boundary indicators       = inner, outer
 # set Zero velocity boundary indicators             = outer
end   


subsection Nullspace removal
   set Remove nullspace = net rotation
end


##### Set temperature Model
subsection Boundary temperature model
#  set Fixed temperature boundary indicators = top, bottom
  set List of model names = spherical constant

  subsection Spherical constant
    set Inner temperature = 3500
    set Outer temperature = 273 #273 for surface ############
  end
end    

#### Temperatur model
subsection Initial temperature model
  #set Model name = adiabatic
  set List of model names =  adiabatic, function
  set List of model operators = add
#  set Model name = adiabatic
  subsection Adiabatic
    set Age bottom boundary layer = 100.e6
    set Age top boundary layer = 100.e6
    set Radius = 0
    set Amplitude = 0
    set Position = center
    set Subadiabaticity = 0
 end 
 subsection Function
     set Coordinate system = cartesian
     set Variable names      = x,y,z
     set Function constants  = R=6371.e3
     set Function expression = \
                               if(sqrt(x*x+y*y)<=R<100e3, rand_seed(1) * 100 -50 , 0)

end
 
subsection Adiabatic conditions model  # need to initialize for initial adiabatic temp.
  set Model name = ascii data
  subsection Ascii data model
    set Data directory = $ASPECT_SOURCE_DIR/data/adiabatic-conditions/ascii-data/
    set Data file name = example_isentrope.txt
  end
end

subsection Heating model
  set List of model names = adiabatic heating, shear heating#, radioactive decay
  subsection  Adiabatic heating
    set Use simplified adiabatic heating = true
  end
end


############################ MATERIAL MODEL ######################################################################

subsection Material model
  set Model name = compositing #visco plastic #melt simple
  set Material averaging = harmonic average
############################ MATERIAL MODEL ######################################################################
  subsection Compositing
    set Viscosity = visco plastic
    set Density = ascii reference profile
    set Thermal expansion coefficient = ascii reference profile
    set Specific heat = ascii reference profile
    set Thermal conductivity = ascii reference profile
    set Compressibility = ascii reference profile
    set Entropy derivative pressure = ascii reference profile
    set Entropy derivative temperature = ascii reference profile
    set Reaction terms = visco plastic
  end
    subsection Ascii reference profile
        set Thermal conductivity =  3.0
        subsection Ascii data model
          set Data file name = example_isentrope.txt
          set Data directory = $ASPECT_SOURCE_DIR/data/adiabatic-conditions/ascii-data/
      end
    end

    subsection Visco Plastic 
     
      set Define transition by depth instead of pressure  = true
      set Phase transition Clapeyron slopes = background: 3.0e6|-2.5e6|9.e6 
      set Phase transition depths           = background: 410e3|660e3|2740e3
      set Phase transition widths           = background: 12e3|23e3|11e3
      set Phase transition temperatures     = background: 1770|1852|2591
      
      # Reference temperature and viscosity
      set Reference temperature = 293
      set Reference viscosity = 6e22
      
      # The minimum strain-rate helps limit large viscosities values that arise
      # as the strain-rate approaches zero.
      # The reference strain-rate is used on the first non-linear iteration
      # of the first time step when the velocity has not been determined yet. 
      set Minimum strain rate = 1.e-20
      set Reference strain rate = 1.e-15

      # Limit the viscosity with minimum and maximum values
      set Minimum viscosity = 1e20
      set Maximum viscosity = 1e24

      # Harmonic viscosity averaging
      set Viscosity averaging scheme = harmonic

      # Choose to have the viscosity (pre-yield) follow a dislocation
      # diffusion or composite flow law.  Here, dislocation is selected
      # so no need to specify diffusion creep parameters below, which are
      # only used if "diffusion" or "composite" option is selected.
      set Viscous flow law = diffusion

#  from cizkov
#     order:                           background    ,        mantle_low        ,        mantle_up
#    set Thermal diffusivities =   8.384146341463415e-07,     8.384146341463415e-07   , 7.309941520467837e-07
   
    set Thermal diffusivities = 1.e-6
    set Thermal expansivities = background: 3.e-5|2.e-5|2.e-5|2.e-5#, plastic_strain: 3.e-5
    set Heat capacities       =        1250     # specific heat
    set Densities             = background: 3416|3800|4750|4750#, plastic_strain: 3416
#    set Thermal expansivities =      3.0e-5  ,              3.0e-5,                     2.7e-5    
#     Dislocation creep                                                 #(1/2)^3.5 * 1e-17 
    set Prefactors for dislocation creep          = background: 1e-22|1e-22|1e-30|1e-30#, plastic_strain:1e-22
    set Stress exponents for dislocation creep    = 3.5
    set Activation energies for dislocation creep = background: 4.8e5|4.8e5|4.8e5|4.8e5#, plastic_strain:4.8e5
    set Activation volumes for dislocation creep  = background: 11.e-6|11.e-6|11.e-6|11.e-6#, plastic_strain: 11.e-6
#     Diffusion creep
    set Prefactors for diffusion creep            = background: 1.e-9|1.e-9|1.0e-16|1.0e-16#, plastic_strain: 1.e-9
    set Stress exponents for diffusion creep      = 1.0                                                  #### doesnt exist anymore??? ####
    set Activation energies for diffusion creep   = background: 3.35e5|3.35e5|2.e5|2.e5#, plastic_strain: 3.35e5
    set Activation volumes for diffusion creep    = background: 4.8e-6|4.8e-6|1.1e-6|1.1e-6#, plastic_strain: 4.8e-6
    set Grain size                                = 1 #1e-3
    set Grain size exponents for diffusion creep  = 0   #default

###    # Plasticity parameters
    set Angles of internal friction = background: 0|0|0|0#, plastic_strain:0            # Angle Mohr
    set Cohesions                   = background: 40e6|1e30|1e30|1e30#, plastic_strain:40e6
   

  end

end

#######################
# Gravity model
subsection Gravity model
  set Model name = ascii data
end

subsection Formulation
  set Formulation = anelastic liquid approximation
end

#######################
# Postprocessing
subsection Postprocess
  set List of postprocessors = velocity statistics, basic statistics, temperature statistics, \
                               visualization, topography, composition statistics, depth average, heat flux statistics, velocity boundary statistics
  subsection Visualization
    set List of output variables = strain rate, stress, nonadiabatic temperature, depth, shear stress, vertical heat flux, heat flux map, material properties, named additional outputs
    set Output format = vtu
    set Time between graphical output = 1e7
    set Interpolate output = true
    set Number of grouped files = 0
    subsection Heat flux map
      set Output point wise heat flux = false  # default=false
    end
  end
  subsection Depth average
    set Number of zones = 10
    set Output format = txt
    set Time between graphical output = 1e7
    set Depth boundaries of zones = 0, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000, 90000, 100000
  end
end
```

MFraters commented:
Hello Elodie,

Thanks for posting the example which didn't work for you. I have tried to reproduce your issue, but unfortunately I was not able to. I tried your input file for both `single Advection, iterated Stokes` and `single Advection, iterated defect correction Stokes` and both methods converge very rapidly for me (see log excerpts below). The defect correction solver converges actually very slightly better, which is in line with expectations. 

The only concern I have is that at timestep zero there is a small difference in the velocity statistics. But after the zero'th timestep, this difference completely disappears. This seems to be due to the fact that the DC Stokes solver only requires two nonlinear iterations, where as the full Stokes solver requires 4 iterations, because when I make the tolerance slightly stricter so that it requires 3 iterations, the result becomes practically the same as the full Stokes solution. This show that it converges to the same solution, so I don't see issue with the forcing terms here. 

Could you post the logs of your runs so that we can compare and see what is not converging in your case, so that we can find out where these differences in results come from?

My results:
full Stokes
```
...cut top part...
*** Timestep 0:  t=0 years, dt=0 years
   Solving temperature system... 0 iterations.
   Rebuilding Stokes preconditioner...
   Solving Stokes system... 81+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 1: 1


Number of active cells: 115,968 (on 9 levels)
Number of degrees of freedom: 1,548,672 (952,320+120,192+476,160)

*** Timestep 0:  t=0 years, dt=0 years
   Solving temperature system... 0 iterations.
   Rebuilding Stokes preconditioner...
   Solving Stokes system... 76+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 1: 1

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 113+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 2: 0.0130744

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 99+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 3: 0.00151489

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 74+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 4: 0.000205122


   Postprocessing:
     RMS, max velocity:                               0.000416 m/year, 0.00223 m/year
     Temperature min/avg/max:                         273 K, 2037 K, 3500 K
     Writing graphical output:                        ellodie-issue-new-stokes/solution/solution-00000
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Writing depth average:                           ellodie-issue-new-stokes/depth_average
     Heat fluxes through boundary parts:              -9.966e+05 W, 1.943e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003565 m/year, 1.105e-07 m/year, 0.0001525 m/year, 3.544e-05 m/year, 2.183e-09 m/year, 1.702e-05 m/year

*** Timestep 1:  t=1e+06 years, dt=1e+06 years
   Solving temperature system... 11 iterations.
   Rebuilding Stokes preconditioner...
   Solving Stokes system... 101+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 1: 0.0645078

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 74+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 2: 0.0201252

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 49+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 3: 0.00166964

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 26+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 4: 0.000178528


   Postprocessing:
     RMS, max velocity:                               0.000431 m/year, 0.00224 m/year
     Temperature min/avg/max:                         273 K, 2037 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.92e+05 W, 1.924e+06 W
     Max, min, and rms velocity along boundary parts: 0.000366 m/year, 2.531e-07 m/year, 0.0001583 m/year, 3.931e-05 m/year, 4.083e-10 m/year, 1.892e-05 m/year

*** Timestep 2:  t=2e+06 years, dt=1e+06 years
   Solving temperature system... 10 iterations.
   Rebuilding Stokes preconditioner...
   Solving Stokes system... 92+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 1: 0.0277684

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 55+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 2: 0.0103391

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 30+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 3: 0.000527453

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 13+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 4: 3.83829e-05


   Postprocessing:
     RMS, max velocity:                               0.000439 m/year, 0.00222 m/year
     Temperature min/avg/max:                         273 K, 2037 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.869e+05 W, 1.908e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003757 m/year, 4.424e-08 m/year, 0.0001627 m/year, 4.145e-05 m/year, 7.137e-11 m/year, 2e-05 m/year

*** Timestep 3:  t=3e+06 years, dt=1e+06 years
   Solving temperature system... 10 iterations.
   Rebuilding Stokes preconditioner...
   Solving Stokes system... 85+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 1: 0.016693

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 50+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 2: 0.00574219

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 25+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 3: 0.000298059


   Postprocessing:
     RMS, max velocity:                               0.000444 m/year, 0.00219 m/year
     Temperature min/avg/max:                         273 K, 2036 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.818e+05 W, 1.899e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003843 m/year, 4.836e-08 m/year, 0.000166 m/year, 4.279e-05 m/year, 3.913e-10 m/year, 2.063e-05 m/year

*** Timestep 4:  t=4e+06 years, dt=1e+06 years
   Solving temperature system... 10 iterations.
   Rebuilding Stokes preconditioner...
   Solving Stokes system... 74+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 1: 0.00935724

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 43+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 2: 0.00208557

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 21+0 iterations.
      Relative nonlinear residual (Stokes system) after nonlinear iteration 3: 0.000132555


   Postprocessing:
     RMS, max velocity:                               0.000446 m/year, 0.00216 m/year
     Temperature min/avg/max:                         273 K, 2036 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.769e+05 W, 1.89e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003917 m/year, 2.216e-07 m/year, 0.0001689 m/year, 4.375e-05 m/year, 7.056e-09 m/year, 2.109e-05 m/year

```

defect correction Stokes:

```
...cut top part...
Number of active cells: 115,968 (on 9 levels)
Number of degrees of freedom: 1,548,672 (952,320+120,192+476,160)

*** Timestep 0:  t=0 years, dt=0 years
   Solving temperature system... 0 iterations.
   Initial Newton Stokes residual = 1.68593e+16, v = 1.68593e+16, p = 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 76+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 0.0450879, norm of the rhs: 7.60153e+14

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 109+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 4.96574e-05, norm of the rhs: 8.37192e+11, newton_derivative_scaling_factor: 0


   Postprocessing:
     RMS, max velocity:                               0.00042 m/year, 0.00225 m/year
     Temperature min/avg/max:                         273 K, 2037 K, 3500 K
     Writing graphical output:                        ellodie-issue-new-dc/solution/solution-00000
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Writing depth average:                           ellodie-issue-new-dc/depth_average
     Heat fluxes through boundary parts:              -9.966e+05 W, 1.943e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003539 m/year, 1.113e-08 m/year, 0.0001531 m/year, 4.014e-05 m/year, 4.432e-10 m/year, 1.933e-05 m/year

*** Timestep 1:  t=1e+06 years, dt=1e+06 years
   Solving temperature system... 11 iterations.
   Initial Newton Stokes residual = 3.47319e+12, v = 3.47319e+12, p = 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 103+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 0.338906, norm of the rhs: 1.17708e+12

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 113+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 0.00505833, norm of the rhs: 1.75685e+10, newton_derivative_scaling_factor: 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 109+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 3: 0.000463673, norm of the rhs: 1.61042e+09, newton_derivative_scaling_factor: 0


   Postprocessing:
     RMS, max velocity:                               0.000431 m/year, 0.00224 m/year
     Temperature min/avg/max:                         273 K, 2037 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.92e+05 W, 1.924e+06 W
     Max, min, and rms velocity along boundary parts: 0.000366 m/year, 2.485e-07 m/year, 0.0001583 m/year, 3.931e-05 m/year, 2.499e-10 m/year, 1.892e-05 m/year

*** Timestep 2:  t=2e+06 years, dt=1e+06 years
   Solving temperature system... 10 iterations.
   Initial Newton Stokes residual = 3.35291e+12, v = 3.35291e+12, p = 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 100+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 0.344216, norm of the rhs: 1.15412e+12

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 112+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 0.004274, norm of the rhs: 1.43303e+10, newton_derivative_scaling_factor: 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 107+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 3: 0.000332627, norm of the rhs: 1.11527e+09, newton_derivative_scaling_factor: 0


   Postprocessing:
     RMS, max velocity:                               0.000439 m/year, 0.00222 m/year
     Temperature min/avg/max:                         273 K, 2037 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.869e+05 W, 1.908e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003757 m/year, 3.853e-08 m/year, 0.0001627 m/year, 4.145e-05 m/year, 8.215e-10 m/year, 2e-05 m/year

*** Timestep 3:  t=3e+06 years, dt=1e+06 years
   Solving temperature system... 10 iterations.
   Initial Newton Stokes residual = 3.17113e+12, v = 3.17113e+12, p = 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 85+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 0.0181053, norm of the rhs: 5.74143e+10

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 105+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 0.000321286, norm of the rhs: 1.01884e+09, newton_derivative_scaling_factor: 0


   Postprocessing:
     RMS, max velocity:                               0.000444 m/year, 0.00219 m/year
     Temperature min/avg/max:                         273 K, 2036 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.818e+05 W, 1.899e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003842 m/year, 4.948e-08 m/year, 0.000166 m/year, 4.278e-05 m/year, 1.868e-09 m/year, 2.063e-05 m/year

*** Timestep 4:  t=4e+06 years, dt=1e+06 years
   Solving temperature system... 10 iterations.
   Initial Newton Stokes residual = 3.14583e+12, v = 3.14583e+12, p = 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 74+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 0.010158, norm of the rhs: 3.19554e+10

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 106+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 0.000140988, norm of the rhs: 4.43524e+08, newton_derivative_scaling_factor: 0


   Postprocessing:
     RMS, max velocity:                               0.000446 m/year, 0.00216 m/year
     Temperature min/avg/max:                         273 K, 2036 K, 3500 K
     Topography min/max:                              -3.725e-09 m, 2.794e-09 m
     Heat fluxes through boundary parts:              -9.769e+05 W, 1.89e+06 W
     Max, min, and rms velocity along boundary parts: 0.0003917 m/year, 2.308e-07 m/year, 0.0001689 m/year, 4.374e-05 m/year, 6.918e-09 m/year, 2.109e-05 m/year

```

elodie-kendall commented:
Hi Menno, thanks for your reply! Actually, the fail happens later. I attach the log files.
For full stokes:
![full_stokes_log](https://user-images.githubusercontent.com/75854106/148746215-a8db8c3e-deba-4033-8b52-7dbf2b855295.png)
Defect CS:
![defect_stokes_log](https://user-images.githubusercontent.com/75854106/148746229-49bfcffa-01ed-4bf6-893b-6cdcb8d7e57b.png)

MFraters commented:
Hey Elodie, Thanks for the clarification and the extra results. 
I don't think we are using exactly the same prm since my results are slightly different from yours, but in both your an my run the results are the same, which is great. What happens with the defect correction solver is that the linear solve fails, but in the results you can see it is very close. Could you run it again with a looser `Linear solver tolerance`. 1e-5 is nearly always overkill for the dc stokes solver and can lead to these kind of issues, 1e-2 or 1e-3 is probably a better default, but 1e-1 will probably also work (it also is a lot faster). The reason for this much looser tolerance is that the full stokes and dc stokes are fundamentally different ways of solving the same problem, and especially the meaning of this tolerance is different. With this I can get past timestep 21 ( t=2.1e+07 years). Let me know if this works for you.

Besides that you could either instead or in addition raise the `Number of cheap Stokes solver steps`

elodie-kendall commented:
Hi Menno, thanks for your speedy response! Ah ok, it is now working with a larger linear tolerance. Thanks so much!

MFraters commented:
Great to hear that it is working for you now! Than I think this pull request is ready for a final review/to be merged.

elodie-kendall commented:
One more issue, now with dislocation creep- this prm below converges for "iterative Stokes" but not for "iterated defect correction Stokes". I've played around with the tolerances and it still doesn't converge. Thanks!

[log.txt](https://github.com/geodynamics/aspect/files/7862363/log.txt)

```
#!/usr/bin/env python3

####  Global parameters
set Dimension                              = 2
set Use years in output instead of seconds = true
set Start time                             = 0
set End time                               = 2000.e6  
set Output directory                       = output
set Nonlinear solver scheme                = single Advection, iterated defect correction Stokes
set Nonlinear solver tolerance             = 5e-4
set Max nonlinear iterations               = 150
set Max nonlinear iterations in pre-refinement = 0
set Pressure normalization                 = surface
set Maximum time step                      = 1e6
#set Number of cheap Stokes solver steps    = 200  # aspect 2.2
set Adiabatic surface temperature          = 1600 
set CFL number                             = 1

set Resume computation          = auto 
subsection Checkpointing
  set Steps between checkpoint  = 100
  set Time between checkpoint   = 0
end
subsection Termination criteria
  set Checkpoint on termination           = true
  set End step                            = 1
  set Termination criteria                = end time #end step
end

subsection Solver parameters
  subsection Stokes solver parameters
    set GMRES solver restart length = 250
    set Linear solver tolerance             = 1e-1
    set Maximum number of expensive Stokes solver steps = 1000
    set Number of cheap Stokes solver steps    = 300 
  end
end

#### Model Geometry
subsection Geometry model
  set Model name = spherical shell

  subsection Spherical shell
    set Inner radius  = 3485e3
    set Outer radius  = 6371e3 #5961e3
    set Opening angle = 360
  end
end

#### Mesh refinement
subsection Mesh refinement
#  set Coarsening fraction = 0.2
#  set Refinement fraction = 0.8
  set Initial adaptive refinement        = 3
  set Initial global refinement          = 5
  set Time steps between mesh refinement = 0
  set Strategy                           = minimum refinement function
  subsection Minimum refinement function
    set Variable names      = depth,t
    set Coordinate system = depth
    set Function expression = if(depth<=200e3, 8,if(depth<=620e3, 7,if(depth<=680e3,6,if(depth>=2586e3,7,5))))
    #if(depth<560e3,0,if(depth>760e3,0,6))
#        if(depth<560e3,5,if(depth>760e3,5,6))
  end
end

#### Set velocity model, zero bottom, tangential others
#subsection Boundary velocity model
#  set Zero velocity boundary indicators       = inner
#  set Tangential velocity boundary indicators = left, right
#end

#subsection Model settings
#  set Fixed composition boundary indicators   = bottom, top 
#  set Fixed temperature boundary indicators   = top, bottom
#  set Zero velocity boundary indicators       = inner
#  set Tangential velocity boundary indicators = left, right
#  #set Prescribed velocity boundary indicators = 3: function
#  set Free surface boundary indicators = top
#end

#subsection Boundary composition model
  #set Allow fixed composition on outflow boundaries = false  # false for models without melt
#  set Fixed composition boundary indicators         = bottom, top
#end
subsection Boundary temperature model
  #set Allow fixed temperature on outflow boundaries = true  # default
  set Fixed temperature boundary indicators         = bottom, top
end
subsection Boundary velocity model
  set Tangential velocity boundary indicators       = inner, outer
 # set Zero velocity boundary indicators             = outer
end   

# Free surface, where exaggeration -> vertical
# subsection Mesh deformation
#   set Mesh deformation boundary indicators        = top: free surface
  #set Free surface boundary indicators = top
#   subsection Free surface
#     set Surface velocity projection             = vertical
#     set Free surface stabilization theta        = 0.5
#   end
# end

subsection Nullspace removal
   set Remove nullspace = net rotation
end

#subsection Mesh deformation
#  set Mesh deformation boundary indicators = top: free surface
#  subsection Free surface
#    set Free surface stabilization theta        = 0.5
#    set Surface velocity projection             = vertical
#  end
#end


# Number and names of compositional fields
#subsection Compositional fields
#  set Number of fields = 1
#  set Names of fields = plastic_strain #add strain field
#end

# Spatial domain of different compositional fields
# Horizontal layers
#subsection Initial composition model
#  set Model name = function
#  subsection Function
#    set Coordinate system    = depth
#    set Variable names       = depth,t                
#    set Function expression  = \
#                               if(depth>=660e3, 1, 0); \
#                               if(depth<660e3, 1, 0)
                               
#subsection Initial composition model
#  set Model name = function
#  subsection Function
#    set Coordinate system    = cartesian
#    set Variable names       = x,y,z 
#    set Function constants   = R=6371.e3               
#    set Function expression  = \
#                               if(sqrt(pow(x,2)+pow(y,2))<660e3,1,0); \
#                               \
#                               if(sqrt(pow(x,2)+pow(y,2))>=660e3,1,0)#,\ 
##                               if((sqrt(pow(x,2)+pow(y,2))>R-200.e3 && \
##                                  sqrt(pow(x,2)+pow(y,2))<R- 20.e3  && \
##                                  atan(y/x)> 139*pi/180 && \
##                                  atan(y/x)< 141*pi/180) || \
##                                  (sqrt(pow(x,2)+pow(y,2))>R-200.e3 && \
##                                  sqrt(pow(x,2)+pow(y,2))<R- 20.e3 && \
##                                  atan(y/x)> 45*pi/180  && \
##                                  atan(y/x)< 46*pi/180)  || \
##                                  (sqrt(pow(x,2)+pow(y,2))>R-200.e3 && \
##                                  sqrt(pow(x,2)+pow(y,2))<R- 20.e3 && \
##                                  atan(y/x)> 88*pi/180  && \
##                                  atan(y/x)< 92*pi/180)    \
##                                  ,0,1)
#                              #if(sqrt(pow(x-4.e6,2)+pow(y-4.e6,2))<600.e3,0,1)#, \
#                                
##                               if(depth<760e3 && depth>=560e3,-(depth-560e3)/(560e3-760e3),if(depth>760e3, 1, 0)); \
##                               if(depth>560e3 && depth<=760e3,-(depth-760e3)/(760e3-560e3),if(depth<560e3, 1, 0))
#                               # Transitional 
##                               if(depth<760e3 && depth>=560e3,-(depth-560e3)/(560e3-760e3),if(depth>760e3, 1, 0)); \
##                               if(depth>560e3 && depth<=760e3,-(depth-760e3)/(760e3-560e3),if(depth<560e3, 1, 0))
#  end
#end

# Composition fixed on bottom, free on sides and top
#subsection Boundary composition model
#  #set Fixed composition boundary indicators = bottom, top 
#  set List of model names = computeprofile # initial compositionS
#  set Model name = initial composition
#end

##### Set temperature Model
subsection Boundary temperature model
#  set Fixed temperature boundary indicators = top, bottom
  set List of model names = spherical constant

  subsection Spherical constant
    set Inner temperature = 4000
    set Outer temperature = 273 #273 for surface ############
  end
end    

#### Temperatur model
subsection Initial temperature model
  #set Model name = adiabatic
  set List of model names =  adiabatic, function
  set List of model operators = add
#  set Model name = adiabatic
  subsection Adiabatic
    set Age bottom boundary layer = 100.e6
    set Age top boundary layer = 100.e6
    set Radius = 0
    set Amplitude = 0
    set Position = center
    set Subadiabaticity = 0
#    subsection Function         
#      set Variable names      = x,y
#      set Function constants  = R=6371.e3
#      set Function expression = \
#                                if(sqrt(x*x+y*y)<=R-660.e3, 1, 0); \
#                                if(sqrt(x*x+y*y)>R-660.e3, 1, 0)
#   end
 end 
 subsection Function
     set Coordinate system = cartesian
     set Variable names      = x,y,z
     set Function constants  = R=6371.e3
     set Function expression = \
                               if(sqrt(x*x+y*y)<=R<100e3, rand_seed(1) * 100 -50 , 0)
 end 

#  subsection Function
#      set Coordinate system = cartesian
#      set Variable names = x,y,z
#      set Function expression = 1
#  end

#  subsection Function
  #  set Coordinate system = depth
  #  set Variable names = depth, t
  #  set Function expression = 1600+0.25*depth/1000
#  end
end
 
subsection Adiabatic conditions model  # need to initialize for initial adiabatic temp.
  #set Model name = compute profile
  #set Model name = initial profile
  #subsection Compute profile #Initial profile
  #    set Variable names      = x,y
  #    set Function constants  = R=6371.e3
  #    set Function expression = \
  #                              if(sqrt(x*x+y*y)<=R-660.e3, 1, 0); \
  #                             if(sqrt(x*x+y*y)>R-660.e3, 1, 0)
#                                if(sqrt(x*x+y*y)>=R-660.e3, 1, 0); \
#                                if(sqrt(x*x+y*y)<R-660.e3, 1, 0)
  #send
  set Model name = ascii data
  subsection Ascii data model
    set Data directory = $ASPECT_SOURCE_DIR/data/adiabatic-conditions/ascii-data/
    set Data file name = example_isentrope.txt
  end
end

subsection Heating model
  set List of model names = adiabatic heating, shear heating, radioactive decay
  subsection  Adiabatic heating
    set Use simplified adiabatic heating = true
  end
  subsection Radioactive decay
    set Crust defined by composition = false 
    set Crust depth = 10e3
    set Half decay times = 4.468e9, 0.704e9, 14e9, 1.248e9
    set Heating rates = 9.4946e-5, 5.68e-4, 2.6368e-5, 2.8761e-5
    set Initial concentrations crust = 0.0, 0.0, 0.0, 0.0
    set Initial concentrations mantle =0.0082, 0.000059, 0.024, 0.013
    set Number of elements = 4
  end
  #set List of model names = compositional heating
end


############################ MATERIAL MODEL ######################################################################

subsection Material model
  set Model name = compositing #visco plastic #melt simple
  set Material averaging = harmonic average
############################ MATERIAL MODEL ######################################################################
  subsection Compositing
    set Viscosity = visco plastic
    set Density = ascii reference profile
    set Thermal expansion coefficient = ascii reference profile
    set Specific heat = ascii reference profile
    set Thermal conductivity = ascii reference profile
    set Compressibility = ascii reference profile
    set Entropy derivative pressure = ascii reference profile
    set Entropy derivative temperature = ascii reference profile
    set Reaction terms = visco plastic
  end
    subsection Ascii reference profile
        set Thermal conductivity =  3.0
        subsection Ascii data model
          set Data file name = example_isentrope.txt
          set Data directory = $ASPECT_SOURCE_DIR/data/adiabatic-conditions/ascii-data/
      end
    end

    subsection Visco Plastic 
     
      set Define transition by depth instead of pressure  = true
      set Phase transition Clapeyron slopes = background: 3.0e6|-2.5e6|9.e6 
      set Phase transition depths           = background: 410e3|660e3|2740e3
      set Phase transition widths           = background: 12e3|23e3|11e3
      set Phase transition temperatures     = background: 1770|1852|2591
      
      # Reference temperature and viscosity
      set Reference temperature = 293
      set Reference viscosity = 6e22
      
      # The minimum strain-rate helps limit large viscosities values that arise
      # as the strain-rate approaches zero.
      # The reference strain-rate is used on the first non-linear iteration
      # of the first time step when the velocity has not been determined yet. 
      set Minimum strain rate = 1.e-20
      set Reference strain rate = 1.e-15

      # Limit the viscosity with minimum and maximum values
      set Minimum viscosity = 1e20
      set Maximum viscosity = 1e24

      # Harmonic viscosity averaging
      set Viscosity averaging scheme = harmonic

      # Choose to have the viscosity (pre-yield) follow a dislocation
      # diffusion or composite flow law.  Here, dislocation is selected
      # so no need to specify diffusion creep parameters below, which are
      # only used if "diffusion" or "composite" option is selected.
      set Viscous flow law = composite # diffusion

#  from cizkov
#     order:                           background    ,        mantle_low        ,        mantle_up
#    set Thermal diffusivities =   8.384146341463415e-07,     8.384146341463415e-07   , 7.309941520467837e-07
   
    set Thermal diffusivities = 1.e-6
    set Thermal expansivities = background: 3.e-5|2.e-5|2.e-5|2.e-5#, plastic_strain: 3.e-5
    set Heat capacities       =        1250     # specific heat
    set Densities             = background: 3416|3800|4750|4750#, plastic_strain: 3416
#    set Thermal expansivities =      3.0e-5  ,              3.0e-5,                     2.7e-5    
#     Dislocation creep                                                 #(1/2)^3.5 * 1e-17 
    set Prefactors for dislocation creep          = background: 4.54e-15|4.54e-15|1e-30|1e-30#, plastic_strain:1e-22
    set Stress exponents for dislocation creep    = 1.0
    set Activation energies for dislocation creep = background: 1.51e5|1.51e5|4.8e5|4.8e5#, plastic_strain:4.8e5
    set Activation volumes for dislocation creep  = background: 17.e-6|17.e-6|11.e-6|11.e-6#, plastic_strain: 11.e-6
#     Diffusion creep
    set Prefactors for diffusion creep            = background: 1.e-9|1.e-9|1.0e-16|1.0e-16#, plastic_strain: 1.e-9
    set Stress exponents for diffusion creep      = 1.0                                                  #### doesnt exist anymore??? ####
    set Activation energies for diffusion creep   = background: 3.35e5|3.35e5|2.e5|2.e5#, plastic_strain: 3.35e5
    set Activation volumes for diffusion creep    = background: 4.8e-6|4.8e-6|1.1e-6|1.1e-6#, plastic_strain: 4.8e-6
    set Grain size                                = 1 #1e-3
    set Grain size exponents for diffusion creep  = 0   #default

###    # Plasticity parameters
    set Angles of internal friction = background: 0|0|0|0#, plastic_strain:0            # Angle Mohr
    set Cohesions                   = background: 40e6|1e30|1e30|1e30#, plastic_strain:40e6
   
    #set Strain weakening mechanism = plastic weakening with plastic strain only
    #set Start plasticity strain weakening intervals = 1.e30
    #set Cohesion strain weakening factors = 1
    #set Friction strain weakening factors = 1
    #set End plasticity strain weakening intervals = 2.e30

  end
################################
  #subsection Composition reaction model

    # Above this depth the compositional fields react: The first field gets
    # converted to the second field. Units: $m$.
   # set Reaction depth                                 = 660.e3

    # Reference density $\rho_0$. Units: $kg/m^3$.
    #set Reference density                              = 3300

    # The value of the specific heat $C_p$. Units: $J/kg/K$.
    #set Reference specific heat                        = 1250

    # The reference temperature $T_0$. Units: $K$.
    #set Reference temperature                          = 293

    # The value of the thermal conductivity $k$. Units: $W/m/K$.
    #set Thermal conductivity                           = 4.7

    # The value of the thermal expansion coefficient $\beta$. Units: $1/K$.
    #set Thermal expansion coefficient                  = 3.0e-5 #Default: 2e-5

    # The temperature dependence of viscosity. Dimensionless exponent.
    #set Thermal viscosity exponent                     = 0.5

    # The value of the constant viscosity. Units: $kg/m/s$.
    #set Viscosity                                      = 1e22 #1e22 #Default: 5e24
    #The value of the width
    #set Reaction width                                 = 5e3 #double
 
    #The value of the smoothing
 #   set Smooth reaction                                = true #bool
 # end  
  
end

#######################
# Gravity model
subsection Gravity model
  set Model name = ascii data
  #set Model name = radial constant
  #subsection Radial constant
  #  set Magnitude = 10
  #end
#  set Model name = ascii data
#  subsection Ascii data model
#    set Data directory = /home/fregehrk/opt/aspect/data/gravity-model/prem.txt
#  end
end

subsection Formulation
  set Formulation = anelastic liquid approximation
end

#######################
# Postprocessing
subsection Postprocess
  set List of postprocessors = velocity statistics, basic statistics, temperature statistics, \
                               visualization, topography, composition statistics, depth average, heat flux statistics
  subsection Visualization
    set List of output variables = strain rate, stress, nonadiabatic temperature, depth, shear stress, vertical heat flux, heat flux map, material properties, named additional outputs
    set Output format = vtu
    set Time between graphical output = 1e7
    set Interpolate output = true
    set Number of grouped files = 0
    subsection Heat flux map
      set Output point wise heat flux = false  # default=false
    end
  end
  subsection Depth average
    set Number of zones = 10
    set Output format = txt
    set Time between graphical output = 1e7
    set Depth boundaries of zones = 0, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000, 90000, 100000
  end
end
```

MFraters commented:
When I try to run this I get the following error:
```
An error occurred in line <43> of file </data/code/aspect/aspect_A/aspect/source/heating_model/radioactive_decay.cc> in function
    void aspect::HeatingModel::RadioactiveDecay<dim>::evaluate(const aspect::MaterialModel::MaterialModelInputs<dim>&, const aspect::MaterialModel::MaterialModelOutputs<dim>&, aspect::HeatingModel::HeatingModelOutputs&) const [with int dim = 2]
The violated condition was: 
    crust_composition_num < material_model_inputs.composition[0].size()
Additional information: 
    The composition number of crust is larger than number of composition
    fields.
```

I see that you are using a different branch than this one. Does that branch include de fixes from this branch?

elodie-kendall commented:
Yes sorry I'm using your branch but with this fix: https://github.com/geodynamics/aspect/pull/4446/files. You can use my branch if it's easier: "menno/fix_compressible_newton_terms" Thanks!

gassmoeller commented on tests/Picard_compressible.prm:
```diff
@@ -0,0 +1,255 @@
+####  Global parameters
```
Please add a comment describing what this tests (same for other test).

gassmoeller reviewed (commented):
Looks good. Please add a changelog entry and see my comment about the tests. After that this is ready to merge.

MFraters commented:
Thanks for the review! I have addressed your points.

gassmoeller commented:
Thanks for the fix!

elodie-kendall commented:
Hi @MFraters! Above you write "So to summarize: I think this fixes both the DC Picard and Newton solver for compressible. The Newton solver for compressible is not stable for this test case and therefore fails." I don't understand, did this PR fix using the newton solver with compressibility too? Thanks!

MFraters commented:
Hey Elodie, 

This pull request makes sure that the nonlinear iteration is solving the correct equations in the compressible case. For DC Picard that is all you need and it should be equal to normal Picard iterations. For Newton iterations there is the chance that the linear system becomes unsolvable. 

A good way of thinking about this figure 10 on page https://aspect-documentation.readthedocs.io/en/latest/user/methods/nonlinear-stokes-solvers.html. The derivative here is zero, so trying to use it to solve the system `sin(x)=0` doesn't work. The linear solve fails because there is no solution to the linear system. For the incompressible Stokes case we found a way to make sure this never happens (note, both the linear and nonlinear solver it can still get into other all kind of other trouble), but in the compressible case some extra terms show up  which make it possible again for this to happen. 

So this pull request makes sure that if the nonlinear iteraion converges, it converges to the correct solution, but there are not the same level of grantees that the linear solver will not fail. It may work, but the linear solver may also fail due to this issue. 

In more technical terms, we want our system to be Symmetric Positive Definite (SPD). I actually build in a way to test if the matrix is SPD in debug mode when you set the velocity block to be SPD (see below). You relatively easily test this because if `A` is SPD, then v^T*(A*v), where v any possible vector, should be positive. So you can just test it by trying lots of `v`'s with random values.

https://github.com/geodynamics/aspect/blob/64c28b2f5267494f19c3c4aa2180d6e07a48a220/source/simulator/assemblers/newton_stokes.cc#L461-L488

I hope his answers some of your questions.

elodie-kendall commented:
I see, thanks for the great explanation @MFraters!
